src/scan/postgres_table_reader.cpp

```diff
@@ -0,0 +1,261 @@
+		if (count_tuples_only) {
+			/* For count_tuples_only we will try to execute aggregate node on table scan */
+			planned_stmt->planTree->parallel_aware = true;
+			marked_parallel_aware = MarkPlanParallelAware((Plan *)table_scan_query_desc->planstate->plan->lefttree);
+		} else {
+			marked_parallel_aware = MarkPlanParallelAware(table_scan_query_desc->planstate->plan);
+		}
+	}
+
+	/* Temp tables can be excuted with parallel workers, and whole plan should be parallel aware */
+	if (persistence != RELPERSISTENCE_TEMP && marked_parallel_aware) {
+
+		int parallel_workers = ParallelWorkerNumber(planned_stmt->planTree->plan_rows);
+		bool interrupts_can_be_process = INTERRUPTS_CAN_BE_PROCESSED();
+
+		if (!interrupts_can_be_process) {
+			RESUME_CANCEL_INTERRUPTS();
+		}
+
+		if (!IsInParallelMode()) {
+			EnterParallelMode();
+			entered_parallel_mode = true;
+		}
+
+		ParallelContext *pcxt;
+		parallel_executor_info = PostgresFunctionGuard(ExecInitParallelPlan, table_scan_planstate,
+		                                               table_scan_query_desc->estate, nullptr, parallel_workers, -1);
+		pcxt = parallel_executor_info->pcxt;
+		PostgresFunctionGuard(LaunchParallelWorkers, pcxt);
+		nworkers_launched = pcxt->nworkers_launched;
+
+		if (pcxt->nworkers_launched > 0) {
+			PostgresFunctionGuard(ExecParallelCreateReaders, parallel_executor_info);
+			nreaders = pcxt->nworkers_launched;
+			parallel_worker_readers = (void **)palloc(nreaders * sizeof(TupleQueueReader *));
+			memcpy(parallel_worker_readers, parallel_executor_info->reader, nreaders * sizeof(TupleQueueReader *));
+		}
+
+		if (!interrupts_can_be_process) {
+			HOLD_CANCEL_INTERRUPTS();
+		}
+	}
+
+	elog(DEBUG1, "(PGDuckdDB/PostgresTableReader)\n\nQUERY: %s\nRUNNING: %s.\nEXECUTING: \n%s", table_scan_query,
+	     !nreaders ? "IN PROCESS THREAD" : psprintf("ON %d PARALLEL WORKER(S)", nreaders),
+	     ExplainScanPlan(table_scan_query_desc).c_str());
+
+	slot = PostgresFunctionGuard(ExecInitExtraTupleSlot, table_scan_query_desc->estate,
+	                             table_scan_planstate->ps_ResultTupleDesc, &TTSOpsMinimalTuple);
+}
+
+PostgresTableReader::~PostgresTableReader() {
+	std::lock_guard<std::mutex> lock(GlobalProcessLock::GetLock());
+
+	PostgresScopedStackReset scoped_stack_reset;
+
+	PostgresFunctionGuard(ExecEndNode, table_scan_planstate);
+
+	if (parallel_executor_info != NULL) {
+		PostgresFunctionGuard(ExecParallelFinish, parallel_executor_info);
+		PostgresFunctionGuard(ExecParallelCleanup, parallel_executor_info);
+	}
+
+	parallel_executor_info = NULL;
+
+	if (parallel_worker_readers) {
+		PostgresFunctionGuard(pfree, parallel_worker_readers);
+	}
+
+	parallel_worker_readers = NULL;
+
+	PostgresFunctionGuard(ExecutorFinish, table_scan_query_desc);
+	PostgresFunctionGuard(ExecutorEnd, table_scan_query_desc);
+	PostgresFunctionGuard(FreeQueryDesc, table_scan_query_desc);
+
+	if (entered_parallel_mode) {
+		ExitParallelMode();
+	}
```

Comment:
It seems that it would be worth moving all that in a `Dispose` private method that would be called here wrapped only once by `PostgresFunctionGuard`?